pep-0544.txt

```diff
@@ -0,0 +1,1006 @@
+where abstractness is simply defined by having at least one abstract method
+being unimplemented. Protocol classes must be marked *explicitly*.
+
+
+Generic and recursive protocols
+-------------------------------
+
+Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
+and ``Iterator`` are generic protocols. They are defined similar to normal
+non-protocol generic types::
+
+  T = TypeVar('T', covariant=True)
+
+  class Iterable(Protocol[T]):
+      @abstractmethod
+      def __iter__(self) -> Iterator[T]:
+          ...
+
+Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``.
+
+Recursive protocols are also supported. Forward references to the protocol
+class names can be given as strings as specified by PEP 484. Recursive
+protocols will be useful for representing self-referential data structures
+like trees in an abstract fashion::
+
+  class Traversable(Protocol):
+      leaves: Iterable['Traversable']
+
+
+Using Protocols
+===============
+
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
+protocol types. For variables and parameters with protocol types, subtyping
+relationships are subject to the following rules:
+
+* A protocol is never a subtype of a concrete type.
+* A concrete type or a protocol ``X`` is a subtype of another protocol ``P``
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
+* Edge case: for recursive protocols, a class is considered a subtype of
+  the protocol in situations where such decision depends on itself.
+  Continuing the previous example::
+
+    class Tree(Generic[T]):
+        def __init__(self, value: T,
+                           leaves: 'List[Tree[T]]') -> None:
+            self.value = value
+            self.leafs = leafs
+
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
+
+Generic protocol types follow the same rules of variance as non-protocol
+types. Protocol types can be used in all contexts where any other types
+can be used, such as in ``Union``, ``ClassVar``, type variables bounds, etc.
+Generic protocols follow the rules for generic abstract classes, except for
+using structural compatibility instead of compatibility defined by
+inheritance relationships.
+
+
+Unions and intersections of protocols
+-------------------------------------
+
+``Union`` of protocol classes behaves the same way as for non-protocol
+classes. For example::
+
+  from typing import Union, Optional, Protocol
+
+  class Exitable(Protocol):
+      def exit(self) -> int:
+          ...
+  class Quitable(Protocol):
+      def quit(self) -> Optional[int]:
+          ...
+
+  def finish(task: Union[Exitable, Quitable]) -> int:
+      ...
+  class GoodJob:
```

Comment:
Can you find a different name for this? "Good job" is such an incredibly boring cliché...